runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectGetWorkStreamTest.java

```diff
@@ -0,0 +1,393 @@
+        .build();
+  }
+
+  @Test
+  public void testSetBudget_computesAndSendsCorrectExtension_noExistingBudget()
+      throws InterruptedException {
+    int expectedRequests = 2;
+    CountDownLatch waitForRequests = new CountDownLatch(expectedRequests);
+    TestGetWorkRequestObserver requestObserver = new TestGetWorkRequestObserver(waitForRequests);
+    GetWorkStreamTestStub testStub = new GetWorkStreamTestStub(requestObserver);
+    stream = createGetWorkStream(testStub, GetWorkBudget.noBudget());
+    GetWorkBudget newBudget = GetWorkBudget.builder().setItems(10).setBytes(10).build();
+    stream.setBudget(newBudget);
+
+    waitForRequests.await(5, TimeUnit.SECONDS);
+
+    // Header and extension.
+    assertThat(requestObserver.sent()).hasSize(expectedRequests);
+    assertThat(Iterables.getLast(requestObserver.sent()).getRequestExtension())
+        .isEqualTo(extension(newBudget));
+  }
+
+  @Test
+  public void testSetBudget_computesAndSendsCorrectExtension_existingBudget()
+      throws InterruptedException {
+    int expectedRequests = 2;
+    CountDownLatch waitForRequests = new CountDownLatch(expectedRequests);
+    TestGetWorkRequestObserver requestObserver = new TestGetWorkRequestObserver(waitForRequests);
+    GetWorkStreamTestStub testStub = new GetWorkStreamTestStub(requestObserver);
+    GetWorkBudget initialBudget = GetWorkBudget.builder().setItems(10).setBytes(10).build();
+    stream = createGetWorkStream(testStub, initialBudget);
+    GetWorkBudget newBudget = GetWorkBudget.builder().setItems(100).setBytes(100).build();
+    stream.setBudget(newBudget);
+    GetWorkBudget diff = newBudget.subtract(initialBudget);
+
+    waitForRequests.await(5, TimeUnit.SECONDS);
+
+    List<Windmill.StreamingGetWorkRequest> requests = requestObserver.sent();
+    // Header and extension.
+    assertThat(requests).hasSize(expectedRequests);
+    assertThat(Iterables.getLast(requests).getRequestExtension()).isEqualTo(extension(diff));
+  }
+
+  @Test
+  public void testSetBudget_doesNotSendExtensionIfOutstandingBudgetHigh()
+      throws InterruptedException {
+    int expectedRequests = 1;
+    CountDownLatch waitForRequests = new CountDownLatch(expectedRequests);
+    TestGetWorkRequestObserver requestObserver = new TestGetWorkRequestObserver(waitForRequests);
+    GetWorkStreamTestStub testStub = new GetWorkStreamTestStub(requestObserver);
+    stream =
+        createGetWorkStream(
+            testStub,
+            GetWorkBudget.builder().setItems(Long.MAX_VALUE).setBytes(Long.MAX_VALUE).build());
+    stream.setBudget(GetWorkBudget.builder().setItems(10).setBytes(10).build());
+
+    waitForRequests.await(5, TimeUnit.SECONDS);
+
+    List<Windmill.StreamingGetWorkRequest> requests = requestObserver.sent();
+    // Assert that the extension was never sent, only the header.
+    assertThat(requests).hasSize(expectedRequests);
+    assertThat(Iterables.getOnlyElement(requests).getRequest())
```

Comment:
this maybe will never succeed since getRequest will return the default object? Or perhaps it throws?

in either case you could just the type of the one-of via assertTrue(...hasRequest()) or something similar. But it could be better to have a matcher that you can use here and other tests that lets you verify the budget within the initial request.

ditto for other instanceof checks

Reply:
done